### PR TITLE
Handle non server action post requests safely

### LINKED ISSUE #56487: transpilePackages causes use of 'import.meta' outside a module, breaking dev mode
### Link to the code that reproduces this issue

https://github.com/sawyerh/nextjs-pino-transpilePackages-bug

### To Reproduce

This bug happens when [`pino`](https://github.com/pinojs/pino) is in the `transpilePackages` config array.

Using the linked repo:

To reproduce the bug, run the application in dev mode:

```
npm run dev
```

You will see the following error:

```
 ⨯ ./node_modules/pino/browser.js
Module parse failed: Cannot use 'import.meta' outside a module (437:16)
|                 // still a Refresh Boundary later.
|                 // @ts-ignore importMeta is replaced in the loader
>                 import.meta.webpackHot.accept();
|                 // This field is set when the previous version of this module was a
|                 // Refresh Boundary, letting us know we need to check for invalidation or
```

To resolve the issue, remove `transpilePackages` from `next.config.js` and restart the server.

### Current vs. Expected behavior

`next dev` fails to run successfully when [`pino`](https://github.com/pinojs/pino) is in the `transpilePackages` config, but `next build && next start` works as expected. I expected `next dev` to also run successfully. 

I am transpiling `pino` in order to support Firefox versions prior to 71.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.6.0: Fri Sep 15 13:41:30 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_ARM64_T8103
Binaries:
  Node: 18.12.1
  npm: 8.19.2
  Yarn: 1.22.19
  pnpm: 8.4.0
Relevant Packages:
  next: 13.5.4
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

SWC transpilation

### Additional context

A similar `import.meta` error message was reported in https://github.com/vercel/next.js/issues/35110 however it's not caused by usage of `transpilePackages`.

<sub>[NEXT-2061](https://linear.app/vercel/issue/NEXT-2061/transpilepackages-causes-use-of-importmeta-outside-a-module-breaking)</sub>

### LINKED ISSUE #58152: Requests with `form-data` or `x-www-form-data-urlencoded` returns error
### Link to the code that reproduces this issue

https://github.com/lkimrxnt/next-formData-issue

#### form-data
1. Start application in development (port 3000)
2. Send `POST` request to the application with any form data:
```
curl --location 'http://localhost:3000/' \
--form 'username="testUserName"' \
--form 'password="testPassword"'
```

### x-www-form-data-urlencoded
1. Start application in development (port 3000)
2. Send `POST` request to the application with any form data:
```
curl --location 'http://localhost:3000/' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'username=testUserName' \
--data-urlencode 'password=testPassword'
```

#### Current
If `form-data` or `x-www-form-data-urlencoded` has been sent the application returns the error below:
```
Missing `origin` header from a forwarded Server Actions request.
 ⨯ TypeError: s is not a function
    at C:\dev\test-next-app\my-app\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:13:5619
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async tE (C:\dev\test-next-app\my-app\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:13:4860)
    at async tV (C:\dev\test-next-app\my-app\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:13:32468)
    at async doRender (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:1294:26)
    at async cacheEntry.responseCache.get.incrementalCache.incrementalCache (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:1446:28)
    at async DevServer.renderToResponseWithComponentsImpl (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:1361:28)
    at async DevServer.renderPageComponent (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:1600:24)
    at async DevServer.renderToResponseImpl (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:1638:32)
    at async DevServer.pipeImpl (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:826:25)
    at async DevServer.handleCatchallRenderRequest (C:\dev\test-next-app\my-app\node_modules\next\dist\server\next-server.js:623:13)
    at async DevServer.handleRequestImpl (C:\dev\test-next-app\my-app\node_modules\next\dist\server\base-server.js:728:17) {
  page: '/'
}
```
If `x-www-form-data-urlencoded` has been sent the equest stuck and never completes

## Expected
The application executes logic in the [middleware](https://github.com/lkimrxnt/next-formData-issue/blob/master/src/middleware.ts), parses formData and renders root page `/`

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

#### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 16.20.0
  npm: N/A
  Yarn: N/A
  pnpm: 8.6.12
Relevant Packages:
  next: 13.5.6
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Middleware / Edge (API routes, runtime)

### Additional context

I've tested the same behavior using the latest version of next (14).

#### Important note
If I send the requests below to a page in `Pages routing` [`/pages/testPage`](https://github.com/lkimrxnt/next-formData-issue/blob/master/src/pages/testPage/index.tsx), everything works as expected.

```
curl --location 'http://localhost:3000/testPage' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'username=testUserName' \
--data-urlencode 'password=testPassword'
```
```
curl --location 'http://localhost:3000/testPage' \
--form 'username="testPassword"' \
--form 'password="testPassword"'
```

<sub>[NEXT-1761](https://linear.app/vercel/issue/NEXT-1761/requests-with-form-data-or-x-www-form-data-urlencoded-returns-error)</sub>

### PR DESCRIPTION
**Pull request by huozhi**
When sending post requests but it's not server action, skip logging warning or calling non-existed server action. Instead we only log the warning like missnig headers for server actions when it's a server action and call the action handler when it's decoded as a function

Fixes #58152 
Closes NEXT-1761

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| buildDuration | 12.8s | 12.8s | N/A |
| buildDurationCached | 7.1s | 6.1s | N/A |
| nodeModulesSize | 199 MB | 199 MB | N/A |
| nextStartRea..uration (ms) | 428ms | 424ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | N/A |
| 433-HASH.js gzip | 28.6 kB | 28.6 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 242 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.2 kB | 45.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| index.html gzip | 529 B | 528 B | N/A |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 148 kB | 148 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 623 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.7 kB | 2.7 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/origin-header-forward | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 169 kB | 169 kB | N/A |
| app-page-exp..prod.js gzip | 95.2 kB | 95.3 kB | ⚠️ +114 B |
| app-page-tur..prod.js gzip | 95.9 kB | 96 kB | ⚠️ +113 B |
| app-page-tur..prod.js gzip | 90.5 kB | 90.6 kB | ⚠️ +111 B |
| app-page.run...dev.js gzip | 142 kB | 142 kB | N/A |
| app-page.run..prod.js gzip | 89.8 kB | 89.9 kB | ⚠️ +109 B |
| app-route-ex...dev.js gzip | 24.1 kB | 24.1 kB | ✓ |
| app-route-ex..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| app-route.ru...dev.js gzip | 23.5 kB | 23.5 kB | ✓ |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.38 kB | 9.38 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.65 kB | 9.65 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.6 kB | 49.5 kB | N/A |
| Overall change | 580 kB | 580 kB | ⚠️ +447 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>433-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5c23e4c71a593ab183ead5469be992039ab8296f

File: packages/next/src/server/app-render/action-handler.ts [hunk 1]
```diff
@@ -498,8 +498,15 @@ export async function handleAction({
             })
             const formData = await fakeRequest.formData()
             const action = await decodeAction(formData, serverModuleMap)
-            const actionReturnedState = await action()
-            formState = await decodeFormState(actionReturnedState, formData)
+            // Only
```
Comment (ztanner) [inline packages/next/src/server/app-render/action-handler.ts:501]:
```suggestion
```

Comment (shuding) [inline packages/next/src/server/app-render/action-handler.ts:501]:
```suggestion
```

File: packages/next/src/server/app-render/action-handler.ts [hunk 2]
```diff
@@ -498,8 +498,15 @@ export async function handleAction({
...
+            if (typeof action === 'function') {
+              // Only warn if it's a server action, otherwise skip for other post requests
+              if (warning) {
+                warn(warning)
```
Comment (ztanner) [inline packages/next/src/server/app-render/action-handler.ts:505]: do we need to handle this in the edge code path as well?

**Review by shuding — APPROVED** [review]
Makes sense, thanks!

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f02c2ac6fa54468feb28652f5cd2bac463733c4b

`pnpm test-dev test/development/acceptance/error-recovery.test.ts` 
- ReactRefreshLogBox turbo > logbox: can recover from a syntax error without losing state

<details>
<summary>Expand output</summary>

  ● ReactRefreshLogBox turbo › logbox: can recover from a syntax error without losing state

    TIMED OUT: /Count: 1/

    Count: 0

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (development/acceptance/error-recovery.test.ts:65:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by ztanner — APPROVED** [review]

**Pull request merged**